### ISSUE TITLE
[RFC]: Phase-aware EOS policy for reasoning and tool-call generation

### ISSUE THREAD

**Issue by QwertyJack** [vllm-project/vllm-ascend#14031]
## Motivation

> **Status:** Future improvement and design discussion. This is not proposed as the immediate fix for #14013.

Reasoning and tool-calling models generate structured output in phases, for example:

```text
reasoning -> reasoning-end -> content or tool call -> EOS
```

Today, vLLM exposes request-wide controls such as `ignore_eos` and `min_tokens`, while reasoning and tool parsers mainly classify tokens after sampling. The existing `thinking_token_budget` machinery can force a reasoning-end sequence when a budget is exhausted, but there is no symmetric, phase-aware policy that can suppress EOS only while an explicitly protected reasoning phase is still open and restore EOS afterwards.

This leaves a gap for model families where EOS is not a valid terminator inside an established reasoning/tool-call protocol. A numerically plausible EOS can terminate generation before the reasoning-end or tool-call marker, producing `finish_reason="stop"` without a tool call.

The DeepSeek V4 investigation in #14013 exposed this as a separate follow-up surface:

- the original Ascend thinking workload produced `7/200` premature-EOS / stop-without-tool-call responses;
- two PA_ND ordering candidates removed physical-allocation-dependent operator output, but still produced `1/21` and `2/34` premature-EOS responses;
- the model-owner-confirmed matching DeepSeek API model completed `100/100` semantically equivalent requests with tool calls and no explicit EOS;
- a temperature-0 official API probe exposed processed top-logprobs for 121 reasoning tokens, and EOS was never returned with a finite logprob.

The official server rendered 56,894 prompt tokens versus 56,815 locally, and the API exposes processed top-logprobs rather than raw full-vocabulary logits. Therefore this evidence does **not** prove that the official implementation masks EOS, nor is it token-exact golden evidence. It does motivate discussing whether vLLM needs an explicit termination-policy abstraction for reasoning/tool protocols.

This RFC is not a proposed fix for the PA_ND allocation-order defect in #14013. Operator repeatability and generation termination policy are separate correctness surfaces.

Upstream context:

- https://github.com/vllm-project/vllm/issues/48645 tracks a DeepSeek V4 parser failure when generation reaches EOS without `</think>`. That is a parser/finalization issue; this RFC concerns whether sampling should permit EOS in an established protected phase.
- vLLM already has `ReasoningConfig`, reasoning start/end token IDs, `ThinkingBudgetStateHolder`, and logits-processor infrastructure. The proposal should extend those generic mechanisms instead of adding an Ascend-only sampler behavior.

## Proposed Change

Introduce an opt-in, phase-aware generation termination policy in upstream vLLM, then validate and integrate it in vLLM Ascend.

### Proposed semantics

Add a policy concept similar to:

```text
premature_eos_policy = "allow" | "mask_in_reasoning"
```

The exact public API is open for discussion. The required semantics are:

1. Default to `allow` for backward compatibility.
2. Derive phase state from actual prompt/generated token sequences and parser metadata, not only from a configured `thinking=true` flag.
3. With `mask_in_reasoning`, mask EOS only after an actual reasoning-start sequence has established an open reasoning phase.
4. Keep reasoning-end and valid tool-start sequences available while EOS is masked.
5. Restore EOS immediately after reasoning-end, an accepted implicit reasoning-end such as tool-start, or another model-declared transition.
6. Preserve `max_tokens` as a hard termination bound.
7. Apply the policy before sampling and top-logprob extraction so returned logprobs describe the effective distribution. Document that they are post-processor logprobs.
8. Define interactions with `min_tokens`, `ignore_eos`, structured output, speculative decoding/MTP, and multi-token phase delimiters.

The policy should be request- or model-configurable and remain opt-in until a model-family contract and regression evidence justify a default. It must not infer that every request with tools requires a tool call: `tool_choice="auto"` can legitimately return content or stop. A stricter tool-required policy, if desired, should be a separate explicit contract.

### Suggested implementation direction

Upstream vLLM:

- extend `ReasoningConfig` with termination-policy metadata;
- reuse or generalize the per-request state tracking in `ThinkingBudgetStateHolder`;
- apply EOS masking in the common logits-processing path for both normal and speculative sampling;
- expose a request-level override through OpenAI-compatible extra fields only after the semantics are agreed;
- keep parser finalization behavior, including https://github.com/vllm-project/vllm/issues/48645, as a separate fallback for outputs that still terminate with a half-open phase.

vLLM Ascend:

- consume the upstream generic policy rather than maintaining an Ascend-specific API;
- ensure the policy is applied consistently in eager/graph, synchronous/async, unified/PD, and MTP paths;
- add DeepSeek V4 A2/A3 regression coverage with BF16 and quantized checkpoints where practical.

### Alternatives considered

**Request-wide `ignore_eos=True`**

Too broad. It also suppresses valid EOS after reasoning/content completion and can force generation to `max_tokens`.

**A large `min_tokens` value**

Length-based rather than protocol-based. It can suppress valid short answers and still permit EOS before a late phase transition.

**Parser-only fallback**

Useful for routing or cleaning a response after termination, as discussed in https://github.com/vllm-project/vllm/issues/48645, but it cannot recover tokens that were never generated after premature EOS.

**Enable the policy automatically whenever thinking is configured**

Unsafe. Some models can answer directly without emitting an explicit reasoning-end marker. Automatic EOS suppression could turn a valid direct answer into a length-limited continuation. Phase establishment and model-family behavior must be explicit.

### Draft validation plan

Unit tests:

- EOS is highest before an established reasoning phase: unchanged under the default policy.
- EOS is highest inside an established protected phase: EOS is masked and a valid continuation/end token is selected.
- reasoning-end or tool-start transitions the request out of the protected phase and restores EOS.
- multi-token reasoning delimiters work across decode steps.
- mixed batches maintain independent phase state.
- returned top-logprobs reflect the post-policy distribution.
- `max_tokens`, `min_tokens`, `ignore_eos`, and structured-output precedence is explicit.

End-to-end tests:

- streaming and non-streaming reasoning plus tool calls;
- normal sampling and MTP/speculative decoding;
- eager and graph execution;
- synchronous and async scheduling;
- unified and PD-separated serving;
- DeepSeek V4 on A2/A3, with BF16 and quantized checkpoints where available;
- direct-answer cases without reasoning delimiters, to prove the opt-in policy does not regress parser fallback behavior.

Success metrics should keep deterministic repeatability separate from task behavior: phase-state correctness, premature-EOS count, tool-call validity, finish reason, output length, and latency overhead should all be reported independently.

## Feedback Period

One week: 2026-08-11 through 2026-08-18.

## Any Other Things

Questions for maintainers:

1. Should the generic abstraction live in `ReasoningConfig`, the logits-processor plugin interface, or a broader generation protocol/state-machine layer?
2. Should the first implementation support only reasoning EOS protection, or also explicit required-terminal policies for tool protocols?
3. Should request-level control use a vLLM extension field, or should it initially be server/model configuration only?
4. Which project should own the first implementation? The proposed preference is upstream vLLM for semantics and vLLM Ascend for backend integration and regression coverage.

**Comment by Zhaocen** [comment]
Implementation is available in #14277.

AI-assisted workflow summary:
- Audited the phase-aware EOS prototype against the vLLM version supported by the current vLLM Ascend main branch.
- Moved the reasoning phase tracker, EOS logits processor, parser transition metadata, and compatibility layer into the Ascend plugin.
- Integrated V1 normal sampling, speculative target verification, and MTP/speculative bonus-token paths.
- Added focused regression coverage for multi-token markers, batch moves/swaps, async placeholders, tool transitions, model-EOS versus user-stop isolation, fail-open behavior, CLI parsing, and clone/msgpack persistence.
- Rebased the single signed commit onto the latest upstream main before opening the PR.
- Targeted unit result: 38 passed. Ruff and available repository hooks passed.
- Real NPU serving was not available in the local macOS environment, so NPU CI validation remains required.
